src/drivers/rover_interface/RoverInterface.cpp

```diff
@@ -0,0 +1,378 @@
+			usleep(5000);
+
+			if (--i == 0) {
+				PX4_ERR("Failed to Stop Task - reboot needed");
+				break;
+			}
+
+		} while (_instance);
+	}
+
+	perf_free(_cycle_perf);
+	perf_free(_interval_perf);
+}
+
+
+int RoverInterface::start(uint8_t rover_type, const char *can_iface)
+{
+	if (_instance != nullptr)
+	{
+		PX4_ERR("Already started");
+		return -1;
+	}
+
+	_instance = new RoverInterface(rover_type, can_iface);
+
+	if (_instance == nullptr)
+	{
+		PX4_ERR("Failed to allocate RoverInterface object");
+		return -1;
+	}
+
+	_instance->ScheduleOnInterval(ScheduleIntervalMs);
+
+	return PX4_OK;
+}
+
+
+void RoverInterface::Init()
+{
+	_initialized = false;
+
+	// Check rover type
+	if (_rover_type == 0)
+	{
+		// Scout Mini
+		PX4_INFO("Scout Mini (rover type 0) is supported. Initializing...");
+	}
+	else if (_rover_type == 1)
+	{
+		// Scout
+		PX4_INFO("Scout (rover type 1) is not supported. Aborted");
+		return;
+	}
+	else if (_rover_type == 2)
+	{
+		// Scout Pro
+		PX4_INFO("Scout Pro (rover type 0) is not supported. Aborted");
+		return;
+	}
+	else if (_rover_type == 3)
+	{
+		// Scout 2
+		PX4_INFO("Scout 2 (rover type 3) is not supported. Aborted");
+		return;
+	}
+	else if (_rover_type == 4)
+	{
+		// Scout 2 Pro
+		PX4_INFO("Scout 2 Pro (rover type 4) is not supported. Aborted");
+		return;
+	}
+	else
+	{
+		// Unknown Rover type
+		PX4_INFO("Unknown rover type. Aborted");
+		return;
+	}
+
+
+	// Check protocol version and create ScoutRobot object
+	if (_protocol_version == scoutsdk::ProtocolVersion::AGX_V1)
+	{
+		PX4_INFO("AGX V1 protocol version is not supported. Aborted");
+	}
+	else if (_protocol_version == scoutsdk::ProtocolVersion::AGX_V2)
+	{
+		PX4_INFO("Detected AGX V2 protocol");
+		_scout = new scoutsdk::ScoutRobot(scoutsdk::ProtocolVersion::AGX_V2, true);
+	}
+	else
+	{
+		PX4_INFO("Unknown protocol version");
+	}
+	if (_scout == nullptr)
+	{
+		PX4_ERR("Failed to create the ScoutRobot object. Aborted");
+		return;
+	}
+
+	// Finally establish connection to rover
+	_scout->Connect(_can_iface);
+	if(!_scout->GetCANConnected())
+	{
+		PX4_ERR("Failed to connect to the rover CAN bus");
+		return;
+	}
+
+	// Use CAN command mode
+	_scout->EnableCommandMode();
+
+	// Retrieve system version (blocking)
+	_scout->QuerySystemVersion(SystemVersionQueryLimitMs);
+
+	// Setup rover state publisher
+	if (!orb_advert_valid(_rover_status_pub))
+	{
+		_rover_status_pub = orb_advertise(ORB_ID(rover_status), &_rover_status_msg);
+	}
+
+	// Breathing mode light by default when not armed
+	_scout->SetLightCommand(LightMode::BREATH, 0);
+
+	_initialized = true;
+}
+
+
+void RoverInterface::Run()
+{
+	pthread_mutex_lock(&_node_mutex);
+
+	if (_instance != nullptr && _task_should_exit.load())
+	{
+		ScheduleClear();
+
+		if (_initialized) { _initialized = false; }
+
+		// Clean up
+		if (_scout != nullptr)
+		{
+			delete _scout;
+			_scout = nullptr;
+		}
+
+		if (_can_iface != nullptr)
+		{
+			free((char *)_can_iface);
```

Comment:
this probably won't work in protected/kernel builds

Reply:
same